### PR TITLE
Rlecomte/tcomp 1696 tck exception model

### PR DESCRIPTION
**Pull request by rlecomte-talend**
https://jira.talendforge.org/browse/TCOMP-1696
### Requirements

This PR allows the component runtime to send a different http reponse code to TDC depending on the source of the error. 520 for any unexpected exception, 400 for exceptions that are set as being caused by a user misconfiguration.

### CODE REVIEW

**Comment by undx** [comment]
2 checkstyle issues: 
```
[ERROR] src/main/java/org/talend/sdk/component/runtime/base/lang/exception/InvocationExceptionWrapper.java:[25,8] (imports) UnusedImports: Unused import - lombok.Getter.
[ERROR] src/main/java/org/talend/sdk/component/runtime/base/lang/exception/InvocationExceptionWrapper.java:[26] (imports) ImportOrder: Wrong order for 'org.talend.sdk.component.api.exception.ComponentException' import.
```

File: component-server-parent/component-server/src/main/java/org/talend/sdk/component/server/front/ActionResourceImpl.java [hunk 1]
```diff
@@ -143,6 +144,18 @@ private Response onError(final Throwable re) {
         if (WebApplicationException.class.isInstance(re.getCause())) {
             return WebApplicationException.class.cast(re.getCause()).getResponse();
         }
+
+        if (ComponentException.class.isInstance(re)) {
+            ComponentException ce = (ComponentException) re;
+            throw new WebApplicationException(Response
+                    .status(ce.getErrorOrigin() == ComponentException.ErrorOrigin.USER ? 400 : 520, "Unexpected callback error")
```
Comment (undx) [inline component-server-parent/component-server/src/main/java/org/talend/sdk/component/server/front/ActionResourceImpl.java:151]: Let's map instead all cases  to :
- ErrorOrigin.USER -> 400
- ErrorOrigin.BACKEND -> 426
- ErrorOrigin.UNKNOWN -> 520

And let handle the 520 case by code below.

File: component-api/src/main/java/org/talend/sdk/component/api/exception/ComponentException.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+ */
+package org.talend.sdk.component.api.exception;
+
+import lombok.Getter;
+
+public class ComponentException extends RuntimeException {
+
+    public static enum ErrorOrigin {
+        USER, // Error caused by user misconfiguration
+        BACKEND, // Error caused by backend
```
Comment (undx) [inline component-api/src/main/java/org/talend/sdk/component/api/exception/ComponentException.java:24]: May prefere either:
- EXECUTION
- RUNTIME
- RUNTIMEEXEC
-...
more than BACKEND.

File: component-server-parent/component-server/src/test/java/org/talend/sdk/component/server/test/custom/CustomService.java [hunk 3]
```diff
@@ -52,4 +55,14 @@ public SuggestionValues get(final LocalConfiguration configuration) throws IOExc
...
+
+    @Action("unknownException")
+    public Map<String, String> generateUnknownException(final LocalConfiguration configuration) {
+        throw new ComponentException(ComponentException.ErrorOrigin.UNKNOWN, "unknown exception");
+    }
+
+    @Action("userException")
+    public Map<String, String> generateUserException(final LocalConfiguration configuration) {
+        throw new ComponentException(ComponentException.ErrorOrigin.USER, "user exception");
+    }
```
Comment (undx) [inline component-server-parent/component-server/src/test/java/org/talend/sdk/component/server/test/custom/CustomService.java:67]: add the missing case also...

File: component-server-parent/component-server/src/main/java/org/talend/sdk/component/server/front/ActionResourceImpl.java (hunk 1 above)
Comment (rlecomte-talend) [reply component-server-parent/component-server/src/main/java/org/talend/sdk/component/server/front/ActionResourceImpl.java:151]: done

File: component-api/src/main/java/org/talend/sdk/component/api/exception/ComponentException.java (hunk 2 above)
Comment (rlecomte-talend) [reply component-api/src/main/java/org/talend/sdk/component/api/exception/ComponentException.java:24]: I use backend to tell the connector could not reach the backend it tries to access to. Just to tell, it's not connector's bad if it failed.

File: component-server-parent/component-server/src/test/java/org/talend/sdk/component/server/test/custom/CustomService.java (hunk 3 above)
Comment (rlecomte-talend) [reply component-server-parent/component-server/src/test/java/org/talend/sdk/component/server/test/custom/CustomService.java:67]: added

File: component-api/src/main/java/org/talend/sdk/component/api/exception/ComponentException.java [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+    }
+
+    public ComponentException(final ErrorOrigin errorOrigin, final String message) {
+        this(errorOrigin, message, null);
+    }
+
+    public ComponentException(final ErrorOrigin errorOrigin, final String message, final Throwable cause) {
+        this(errorOrigin, cause != null ? cause.getClass().getName() : null, message, null, cause);
+    }
+
```
[the remark this answers is not in the document]
Comment (rlecomte-talend) [reply component-api/src/main/java/org/talend/sdk/component/api/exception/ComponentException.java:60]: done

**Comment by undx** [comment]
Still one checkstyle error:
```[ERROR] src/main/java/org/talend/sdk/component/runtime/base/lang/exception/InvocationExceptionWrapper.java:[20] (imports) ImportOrder: Wrong order for 'java.lang.reflect.InvocationTargetException' import.```

**Pull request merged**